target/linux/realtek/dts-5.10/rtl8393_tplink_sg2452p-v4.dts

```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "rtl839x.dtsi"
+
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	compatible = "tplink,sg2452p-v4", "realtek,rtl838x-soc";
+	model = "TP-Link SG2452P v4";
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x10000000>;
+	};
+
+	aliases {
+		led-boot = &led_sys;
+		led-failsafe = &led_sys;
+		led-running = &led_sys;
+		led-upgrade = &led_sys;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,38400";
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			gpios = <&gpio0 20 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
+		speed {
+			label = "speed";
+			gpios = <&gpio0 19 GPIO_ACTIVE_LOW>;
+			linux,code = <BTN_0>;
+		};
+	};
+
+	gpio_fan_sys {
+		compatible = "gpio-fan";
+		alarm-gpios = <&gpio0 12 GPIO_ACTIVE_LOW>;
+	};
+
+	gpio_fan_psu_1 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&disable_jtag>;
+		compatible = "gpio-fan";
+
+		alarm-gpios = <&gpio0 7 GPIO_ACTIVE_LOW>;
+		gpios = <&gpio0 4 GPIO_ACTIVE_LOW>;
+		/* the actual speeds (rpm) are unknown, just use dummy values */
+		gpio-fan,speed-map = <1 0>, <2 1>;
+		#cooling-cells = <2>;
+	};
+
+	gpio_fan_psu_2 {
+		compatible = "gpio-fan";
+		alarm-gpios = <&gpio0 14 GPIO_ACTIVE_LOW>;
```

Comment:
Same as for `gpio_fan_sys`: does this fan run at a fixed speed, or is this a work-around for `alarm-gpios` on `gpio_fan_psu_1` only supporting one GPIO line?

Reply:
Exactly - both PSU fans can be slowed down by the same GPIO line but have individual alarm lines.

Reply:
If all fans are controlled by the same set of GPIOs, then I think the devicetree description should be able to reflect this.

I think we could propose upstream to use [`devm_gpiod_get_array_optional()`](https://elixir.bootlin.com/linux/latest/source/include/linux/gpio/consumer.h#L102) to support multiple alarm lines for `alarm-gpios`. That call is also available in 5.10, so the change could probably be backported.

Reply:
Sounds like an option, but after my recent submission regarding the tps23861 driver I will not get in touch with anything kernel related in the near future.

Reply:
Upstream can be hard :-/

I've started with converting the binding to YAML and may have a look at multi-alarm-gpios after next week. We can keep the DT like you've suggested for now, although it should be documented that both fans are controlled by the GPIO values.

Reply:
@andyboeh You cannot really think like that, upstreaming is hard and won't get easier, but is the only way to prevent maintaining local patches forever.

OpenWrt is full of never upstreamed hacks that have piled up over the years as people treated it way around dealing with upstream, you can probably see what an issue every kernel bump is.

Just look at the mess of a target Realtek is, its still stuck on 5.10

Reply:
I know and I'm not against upstreaming, don't get me wrong here. It's just that I do not have the time and resources to do that. That's also one of the reasons why I deleted the tps23861 patches instead of including a local copy here.

It's hard enough for me to contribute to OpenWrt at all - I wanted a port to this switch model since I bought it in 2020 and just now I have a bit of time to do so. However, this period is limited and will be over or at least very limited in two weeks due to work and family obligations.

I'm very glad that people like you contribute much more to this project than I can do.

Reply:
We'll see how far we can get in that short time span then. Just try to document every little detail you know, that could be very helpful if others want to write changes when drivers are updated to provide more features.

Reply:
I understand time is usually the limiting factor anywhere.
Though, it looks like PoE PSE basic framework is slowly getting assembled so in future we could add 802.3at/af/bt support as well instead of having multiple userspace implementations.